Implementation/59288 add stages and gates to the project overview page

dombesz opened the pull request:
# Ticket
https://community.openproject.org/wp/59288

# What are you trying to accomplish?
Display Stages and Gates on the Project Overview page

## Screenshots

# What approach did you choose and why?
Similar implementation to ProjectCustomFields, as there are a lot of similarities.

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [x] Tested major browsers (Chrome, Firefox, Edge, ...)

dombesz committed:
wip

dombesz committed:
Display life cycles in the sidebar

dombesz committed:
Update variable names

dombesz committed:
Add Stages and Gates permissions

dombesz committed:
Allow instantiation of Project::LifeCycleStepDefinition in order to let acts_as_list function correctly.

dombesz committed:
Rename overview page spec helper methods

dombesz committed:
Add sidebar specs for displaying life cycles

dombesz committed:
Add view permission specs

dombesz committed:
Add edit permission specs

dombesz committed:
Add Coloured icons, initial rendering of the form fields

dombesz committed:
Add DatePicker component and dsl for single_date_picker and range_date_picker form fields

dombesz committed:
Display the datepicker input fields on the life cycle edit form

dombesz committed:
Add ProjecLifeCycleSteps::BaseContract and UpdateContract.

dombesz committed:
Make form validations work for nested project life cycle attributes.

ulferts committed:
exclude OP custom dom elements from being morphed

dombesz committed:
Add Project::Stage date_range setter and validation.

dombesz committed:
Remove condition for invalid LifeCycleStep initialization, add validation instead.

dombesz committed:
Use date and date_range in the life cycle steps dialog, automatically infer value in the datepicker, complete the dialog workflow.

dombesz committed:
Add life cycle step increasing dates validation to the ProjectLifeCycleStep::BaseContract

dombesz committed:
Activate associated validations for project life cycle steps only when validating via the ProjectLifeCycleSteps contracts.

dombesz committed:
Display duration in days caption for Project::Stage entries.

dombesz committed:
Create form preview to interactively calculate duration of selected dates and display error messages on the ProjectLifeCycle dialog.

dombesz committed:
Allow empty date and date range for Project::LifeCycleStep

dombesz committed:
Enable morphing the Stages and Gates form preview action

dombesz committed:
Add validation error border around the datepicker's input.

dombesz committed:
Add specs for stages and gates edit dialog

dombesz committed:
Addd more Stage and Gates dialog specs wip.

dombesz committed:
Fix unit specs

EinLama commented on config/locales/en.yml:
```diff
@@ -1186,16 +1189,38 @@ en:
+          attributes:
+            date_range:
+              start_date_must_be_before_end_date: "start date must be before the end date."
+              non_continuous_dates: "can’t be earlier than the previous %{step}'s end date."
```
same as above: nitpick: two different styles for `'` in this sentence. `can't` has a different apostrophe than `%{step}'s`. Is this by choice or chance? AFAIK it should be `'` everywhere.

EinLama commented on config/locales/en.yml:
```diff
@@ -1186,16 +1189,38 @@ en:
+            type:
+              type_and_class_name_mismatch: "must be a Project::Gate"
+            date:
+              non_continuous_dates: "can’t be earlier than the previous %{step}'s end date."
```
nitpick: two different styles for `'` in this sentence. `can't` has a different apostrophe than `%{step}'s`. Is this by choice or chance? AFAIK it should be `'` everywhere.

EinLama commented on app/models/project/stage.rb:
```diff
@@ -29,5 +29,36 @@
+  end
+
+  def range_incomplete?
+    start_date.blank? ^ end_date.blank?
```
Clever use of XOR 👍🏻

EinLama commented on app/models/project/stage.rb:
```diff
@@ -29,5 +29,36 @@
+  def working_days_count
+    return nil if not_set?
+
+    Day.working.from_range(from: start_date, to: end_date).count
```
This would be much more complicated if we had the option of setting work-week-days per project (or system-wide) 😮

EinLama commented on frontend/src/turbo/turbo-global-listeners.ts:
```diff
@@ -64,4 +64,13 @@ export function addTurboGlobalListeners() {
+    const element = event.target as HTMLElement;
+
+    // In case the element is an OpenProject custom dom element, morphing is prevented.
+    if (element.tagName.startsWith('OPCE-')) {
```
If the HTML spec ever changes, we encounter a wonky Browser, or if we ever migrate to XHTML, the tagName might not be all-uppercase, see [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Element/tagName#examples). While unlikely, I suggest ruling out this hard-to-debug edge-case and safeguard this condition by always converting the tagName to uppercase, just in case.

```suggestion
    if (element.tagName.toUpperCase().startsWith('OPCE-')) {
```

EinLama commented on lib/primer/open_project/forms/date_picker.html.erb:
```diff
@@ -0,0 +1,24 @@
+<%= render(FormControl.new(input: @input, tag: :"primer-datepicker-field")) do %>
```
Consider adding the Copyright comment above here.

EinLama commented on lib/primer/open_project/forms/date_picker.rb:
```diff
@@ -0,0 +1,19 @@
+# frozen_string_literal: true
```
Consider adding the Copyright comment here :)

EinLama commented on lib/primer/open_project/forms/dsl/range_date_picker_input.rb:
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
```
Consider adding the Copyright comment here :)

EinLama commented on lib/primer/open_project/forms/dsl/single_date_picker_input.rb:
```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
```
Consider adding the Copyright comment here :)

EinLama commented on modules/overviews/app/components/project_life_cycles/sections/edit_component.rb:
```diff
@@ -0,0 +1,41 @@
+      include OpPrimer::ComponentHelpers
+
+      def life_cycle_steps
+        model.life_cycle_steps.active.eager_load(:definition).order(position: :asc)
```
You could re-use previous logic here:

```suggestion
        model.available_life_cycle_steps
```

EinLama commented on modules/overviews/app/components/project_life_cycles/sections/edit_dialog_component.html.erb:
```diff
@@ -0,0 +1,32 @@
+      component_collection do |footer_collection|
+        footer_collection.with_component(Primer::ButtonComponent.new(
+          data: {
+            'close-dialog-id': "edit-project-life-cycles-dialog"
```
Since this ID refers to the Dialog id defined in line 4, it would make sense to move this into a constant/method within the component. That way, if the ID changes in the future, you only need to change it in one place.

e.g.

```ruby
module ProjectLifeCycles
  module Sections
    class EditComponent
       def dialog_id
         "edit-project-life-cycles-dialog"
       end
# ...

# here:
'close-dialog-id': dialog_id
```

EinLama commented on modules/overviews/app/components/project_life_cycles/sections/project_life_cycles/show_component.rb:
```diff
@@ -0,0 +1,79 @@
+          model.not_set?
+        end
+
+        def render_value
+          case model
+          when Project::Gate
+            render(Primer::Beta::Text.new) do
+              concat helpers.format_date(model.date)
+            end
+          when Project::Stage
+            render(Primer::Beta::Text.new) do
+              concat [
+                helpers.format_date(model.start_date),
+                helpers.format_date(model.end_date)
+              ].join(" - ")
+            end
+          end
+        end
```
This could be simplified so that you don't have to check which type the model is.

```suggestion
        def render_value
          render (Primer::Beta::Text.new) do
            concat [
              model.start_date,
              model.end_date
            ]
            .compact
            .map { |d| helpers.format_date(d) }
            .join(" - ")
          end
        end
```

EinLama commented on modules/overviews/app/components/project_life_cycles/sections/show_component.rb:
```diff
@@ -0,0 +1,51 @@
+        super
+
+        @project = project
+        @life_cycle_steps =
+          @project.life_cycle_steps.active.eager_load(:definition).order(position: :asc)
```
```suggestion
        @life_cycle_steps = @project.available_life_cycle_steps
```

EinLama commented on modules/overviews/app/components/project_life_cycles/sections/show_component.html.erb:
```diff
@@ -0,0 +1,34 @@
+
+      flex_layout do |details_container|
+        @life_cycle_steps.each_with_index do |life_cycle_step, i|
+          margin =  i == @life_cycle_steps.size - 1 ? 0 : 3
```
bonus space

```suggestion
          margin = i == @life_cycle_steps.size - 1 ? 0 : 3
```

EinLama commented on modules/overviews/app/controllers/overviews/overviews_controller.rb:
```diff
@@ -49,6 +76,50 @@ def update_project_custom_values
+        component: ProjectLifeCycles::Sections::EditComponent.new(service_call.result),
+        method: "morph"
+      )
+      # TODO: :unprocessable_entity is not nice, change the dialog logic to accept :ok
+      # without dismissing the dialog, alternatively use turbo frames instead of streams.
```
While true, I think it is safe to say that this is outside of the scope of this PR 👍🏻

EinLama commented on modules/overviews/lib/overviews/engine.rb:
```diff
@@ -47,7 +47,8 @@ class Engine < ::Rails::Engine
           .push(
-            "overviews/overviews/show"
+            "overviews/overviews/show",
+            "overviews/overviews/project_life_cycles_sidebar"
```
It surprised me that the side bar is part of `view_project`. If I understand it correctly, you must have both `view_project` and `view_project_stages_and_gates` permissions to see the sidebar.

Isn't this superfluous as you wouldn't see the entire overview page to begin with if you lacked the `view_project` permission? 🤔 

In other words: is it necessary to add this line?

dombesz replied:
Yep, it is superfluous, I think it is just a typo. Thanks!

EinLama commented on spec/contracts/project_life_cycle_steps/base_contract_spec.rb:
```diff
@@ -0,0 +1,153 @@
+  let(:contract) { described_class.new(project, user) }
+  let(:project) { build_stubbed(:project) }
+
+  context "with authorised user" do
```
Americanize:

```suggestion
  context "with authorized user" do
```

EinLama commented on spec/contracts/project_life_cycle_steps/base_contract_spec.rb:
```diff
@@ -0,0 +1,153 @@
+    end
+  end
+
+  context "with unauthorised user" do
```
```suggestion
  context "with unauthorized user" do
```

dombesz committed:
Attempt to fix flaky specs.

EinLama commented on spec/models/project/stage_spec.rb:
```diff
@@ -33,13 +33,112 @@
+      expect(subject.working_days_count).to eq(3)
+    end
+
+    it "calls the Day.working.from_range method with the right arguments" do
+      subject.start_date = Date.parse("2024-11-25")
+      subject.end_date = Date.parse("2024-11-27")
+
+      allow(Day).to receive(:working).and_return(Day)
+      allow(Day).to receive(:from_range)
+                .with(from: subject.start_date, to: subject.end_date)
+                .and_return([])
+
+      expect(subject.working_days_count).to eq(0)
+    end
```
Can you explain why you added this spec?

This spec is a bit odd as it tests the internals of the method in question. I wonder if it wouldn't be better to just supply some more test cases with specific date ranges and expected amounts of working days?

Right now, this spec would break if the `working_days_count` method changes its implementation, even if the result returned was the same.

dombesz replied:
I agree, the tests became a bit more brittle having this expectation. However, I believe the costs weights the benefits. AFAIK it is ok to mock external objects that are independent from the class, because I'm not mocking the internal methods of this class.

This way I can make sure the right behaviour is called, without duplicating all the tests from the `day_spec.rb`.

EinLama replied:
I see, that makes sense.

With that in mind, we should check if the allowed methods were actually called:
```ruby
      expect(Day).to have_received(:working).with(no_args)
      expect(Day).to have_received(:from_range).with(from: subject.start_date, to: subject.end_date)
```

dombesz replied:
Very good point, for some reason that was left off 😅

EinLama commented on spec/support/pages/projects/show.rb:
```diff
@@ -56,7 +54,12 @@ def expect_no_visible_sidebar
-      def within_async_loaded_sidebar(&)
+      def expect_visible_sidebar
+        expect_angular_frontend_initialized
+        expect(page).to have_css(".op-grid-page--grid-container")
```
I think the CSS class here is wrong 😁 shouldn't it rather be:

```suggestion
        expect(page).to have_css(".op-grid-page--sidebar")
```

EinLama reviewed (commented):
Works nicely! I have added some remarks, most of which are small things. I could not find any major flaws, great job!

dombesz committed:
Use the correct Project Lifecycle label

dombesz commented:
Thanks @EinLama  for the quick review, I believe I addressed all your comments.

dombesz committed:
Address CR comments

dombesz committed:
Fix specs

dombesz committed:
Fix single quote character

dombesz committed:
Allow retry_block on RSpec::Expectations::ExpectationNotMetError too.

This exception is not included by default, because it does not inherit
from the StandarError, but from the Exception.

dombesz committed:
Fix rubocop issues

EinLama commented on spec/features/projects/life_cycle/overview_page/dialog/permission_spec.rb:
```diff
@@ -0,0 +1,84 @@
+  let(:permissions) { [] }
+
+  before do
+    allow(User).to receive(:current).and_return user
```
You could also use the `current_user` [helper](https://github.com/opf/openproject/blob/1d73cab4a8932ccf8ecd40ab4a81f52ddb54313c/spec/support/authentication_helpers.rb#L104) here:

```ruby
current_user { user }
```

EinLama commented on spec/features/projects/life_cycle/overview_page/dialog/update_spec.rb:
```diff
@@ -0,0 +1,155 @@
+
+  before do
+    # TODO: Could this work for all feature specs?
+    allow(User).to receive(:current).and_return admin
```
You could also use the `current_user` [helper](https://github.com/opf/openproject/blob/1d73cab4a8932ccf8ecd40ab4a81f52ddb54313c/spec/support/authentication_helpers.rb#L104) here:

```ruby
current_user { admin }
```

EinLama commented on spec/features/projects/life_cycle/overview_page/sidebar_spec.rb:
```diff
@@ -0,0 +1,153 @@
+    it "shows the project stages and gates in the correct order" do
+      overview_page.visit_page
+
+      overview_page.within_life_cycles_sidebar do
+        fields = page.all(".op-project-life-cycle-container")
+
+        expect(fields.size).to eq(7)
+
+        expect(fields[0].text).to include("Initiating")
+        expect(fields[1].text).to include("Ready for Planning")
+        expect(fields[2].text).to include("Planning")
+        expect(fields[3].text).to include("Ready for Executing")
+        expect(fields[4].text).to include("Executing")
+        expect(fields[5].text).to include("Ready for Closing")
+        expect(fields[6].text).to include("Closing")
+      end
```
This could be simplified:

```suggestion
    overview_page.within_life_cycles_sidebar do
      expected_stages = [
        "Initiating",
        "Ready for Planning",
        "Planning",
        "Ready for Executing",
        "Executing",
        "Ready for Closing",
        "Closing"
      ]

      fields = page.all(".op-project-life-cycle-container")
      expect(fields.map(&:text)).to eq(expected_stages)
    end
```

EinLama commented on spec/features/projects/life_cycle/overview_page/sidebar_spec.rb:
```diff
@@ -0,0 +1,153 @@
+
+        expect(fields.size).to eq(7)
+
+        expect(fields[0].text).to include("Initiating")
```
Same simplification possible here :)

EinLama reviewed (approved):
Nice one 👏🏻

dombesz committed:
Address CR comments
